### PR TITLE
Closes #5091 Added RUCSS exclusion for Polygone

### LINKED ISSUE #5091: RUCSS compatibility with Polygone theme expanding sections
RUCSS compatibility with Polygone theme expanding sections

**Before submitting an issue please check that you’ve completed the following steps:**
- [x] Made sure you’re on the latest version `3.11.2`
- [x] Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
Polygone theme (child theme for Divi) has a feature to expand section when they reach the middle of the viewport. 

That particular section has a dynamic selector that changes on each page cache refresh.

```
<script>
    jQuery(document).ready(function(){
        jQuery('.et_pb_b3_promo.expanding_bar_392').parent().parent().parent().addClass('et_pb_expanding_bar_section expanding_bar_392');
        jQuery('.et_pb_b3_promo.expanding_bar_392').parent().parent().addClass('et_pb_row_fullwidth');	
        jQuery('body').addClass('expanding_bar_highlight');
        jQuery('.et_pb_expanding_bar_section.expanding_bar_392').waypoint(function() {
            jQuery('body').toggleClass('expanding_bar_392');
            jQuery('body').toggleClass('poly_expanding_bar_highlight');
            jQuery('.et_pb_expanding_bar_section.expanding_bar_392 .et_pb_b3_promo_button').toggleClass('expanding_bar_highlight_button');
            jQuery('.et_pb_expanding_bar_section.expanding_bar_392 .et_pb_b3_promo_description').toggleClass('expanding_bar_highlight_description');
        }, {offset: '50%'});
    });
</script>
<style>
.expanding_bar_highlight.expanding_bar_392 .et_pb_expanding_bar_section.expanding_bar_392 .et_pb_row.et_pb_row_fullwidth{ 
    width: 100% !important;
    max-width:100% !important;}
    .et_pb_expanding_bar_section.expanding_bar_392 .et_pb_row.et_pb_row_fullwidth{background-image: url(https://domain.ext/wp-content/uploads/2017/01/marco-mancosu-contact-me.jpg) !important;background-position: initial;} 
</style>
```

It's added as following on the code:

```
$rand =  "expanding_bar_".rand(10,500);
'selector'    => '.et_pb_expanding_bar_section.'.$rand.'
```

[/wp-content/themes/Polygon/b3_expanding_bar_module.php](https://gist.github.com/DahmaniAdame/f5b7437a93ff2a2642ec80eefca08a58)

Adding `.expanding_bar_` to `rocket_rucss_inline_content_exclusions` fixes the issue.

**To Reproduce**
1. Use Polygon theme and add an expandable section
2. Use RUCSS
3. See error - the supposed to be expanded section is missing its style. 

**Expected behavior**
Expandable sections should work out of the box.

**Screenshots**
N/A

**Additional context**
Ticket - https://secure.helpscout.net/conversation/1897304223/345070/

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description
Created a third party class for Polygone that excludes `.expanding_bar_` from RUCSS when the theme is active.

Fixes #5091

## Type of change

- [ ] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

- [ ] Automated Tests

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/Plugin.php [hunk 1]
```diff
@@ -275,6 +275,7 @@ private function init_common_subscribers() {
 			'rucss_frontend_subscriber',
 			'rucss_cron_subscriber',
 			'divi',
+			'polyone',
```
Comment (jeawhanlee) [inline inc/Plugin.php]:
```suggestion
			'polygone',
```

File: inc/ThirdParty/ServiceProvider.php [hunk 2]
```diff
@@ -88,6 +89,9 @@ public function register() {
 			->addArgument( $options )
 			->addArgument( $this->getContainer()->get( 'delay_js_html' ) )
 			->addTag( 'common_subscriber' );
+		$this->getContainer()
+			->share( 'polyone', 'WP_Rocket\ThirdParty\Themes\Polygone' )
```
Comment (jeawhanlee) [inline inc/ThirdParty/ServiceProvider.php]:
```suggestion
			->share( 'polygone', 'WP_Rocket\ThirdParty\Themes\Polygone' )
```

**Review by mostafa-hisham — APPROVED** [review]

**Comment by vmanthos** [comment]
@CrochetFeve0251 I tested this and it's not working. This is because the wrong filter is used here:
https://github.com/wp-media/wp-rocket/blob/a4fc57c81a629330da5269bcc4e7431c5f362267/inc/ThirdParty/Themes/Polygone.php#L20

It should be this one instead:
`rocket_rucss_inline_content_exclusions`
https://github.com/wp-media/wp-rocket/blob/e301cb1503f14bafd3c0f16cafb0233910d7618d/inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php#L432

Can you please make the change?

**Comment by vmanthos** [comment]
@CrochetFeve0251 Thank you for the PR.

There is a syntax error here:
https://github.com/wp-media/wp-rocket/blob/83e4fa20acd66da5a078eaf336101ae9e52bf023/inc/ThirdParty/ServiceProvider.php#L38

The line should end with a `;` instead of a `:`.

**Comment by CrochetFeve0251** [comment]
> @CrochetFeve0251 I tested this and it's not working. This is because the wrong filter is used here:
> 
> https://github.com/wp-media/wp-rocket/blob/a4fc57c81a629330da5269bcc4e7431c5f362267/inc/ThirdParty/Themes/Polygone.php#L20
> 
> It should be this one instead: `rocket_rucss_inline_content_exclusions`
> 
> https://github.com/wp-media/wp-rocket/blob/e301cb1503f14bafd3c0f16cafb0233910d7618d/inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php#L432
> 
> Can you please make the change?

@vmanthos I fixed it and that's possible as if I remember well  I don't have Poygone theme to test directly on it.

**Review by vmanthos — COMMENTED** [review]
@CrochetFeve0251 This is working now. Thank you for the quick fix.

I have a couple of notes for you to consider, please.

1. Currently, this will work if the theme's first letter is capital, i.e. `Polygone`, but not when it's `polygone`. 

What about making this case insensitive? 

2. The original issue is about the [Polygone theme](https://www.b3multimedia.ie/product/polygon/) which is a child theme for Divi. 

The way this is currently coded if a `parent` theme named `Polygone` is installed/activated our code will run unnecessarily. 

How about us being more explicit about checking for Divi's Polygone child theme here?
https://github.com/wp-media/wp-rocket/blob/083123c855ec89dad65c5b940833d8b4ae41f8e1/inc/ThirdParty/Themes/Polygone.php#L42

**Comment by CrochetFeve0251** [comment]
> @CrochetFeve0251 This is working now. Thank you for the quick fix.
> 
> I have a couple of notes for you to consider, please.
> 
>     1. Currently, this will work if the theme's first letter is capital, i.e. `Polygone`, but not when it's `polygone`.
> 
> 
> What about making this case insensitive?
> 
>     2. The original issue is about the [Polygone theme](https://www.b3multimedia.ie/product/polygon/) which is a child theme for Divi.
> 
> 
> The way this is currently coded if a `parent` theme named `Polygone` is installed/activated our code will run unnecessarily.
> 
> How about us being more explicit about checking for Divi's Polygone child theme here?
> 
> https://github.com/wp-media/wp-rocket/blob/083123c855ec89dad65c5b940833d8b4ae41f8e1/inc/ThirdParty/Themes/Polygone.php#L42

Concerning point 1 I will do it.

Concerning point 2 we need to exclude also child themes as most of agencies I worked with none used the theme directly but all of them created a child theme from the theme they chose to had modifications required by the client.

**Comment by vmanthos** [comment]
> I don't have Poygone theme to test directly on it

I've created a "dummy" child theme based on how that's set on the customer's website:
https://github.com/wp-media/rocket-test-data/tree/master/themes/Polygone

You can use the following template:
https://github.com/wp-media/rocket-test-data/blob/master/templates/rucss_5091_template.php

 with our [Template Loader plugin](https://github.com/wp-media/template-loader-plugin)

@DahmaniAdame Would it be possible to ask the theme from the customer so we can test the fix with that as well, please?

**Comment by vmanthos** [comment]
@CrochetFeve0251 The theme's name is `Polygon` without an `e`:
https://marcomancosu.com/wp-content/themes/Polygon/style.css?ver=4.17.4

I had it wrong in my "dummy" theme as well. It's fixed now:
https://github.com/wp-media/rocket-test-data/tree/master/themes/Polygon

Can you please change this? 🙏

**Comment by vmanthos** [comment]
@CrochetFeve0251 With the latest commit I'm getting the following fatal error:

```PHP
[08-Jul-2022 05:31:47 UTC] PHP Fatal error:  Uncaught WP_Rocket\Dependencies\League\Container\Exception\ContainerException: Service provider lied about providing (polygon) service in /var/www/example.com/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php:175
Stack trace:
#0 /var/www/example.com/htdocs/wp-content/plugins/wp-rocket/inc/Plugin.php(143): WP_Rocket\Dependencies\League\Container\Container->get()
#1 /var/www/example.com/htdocs/wp-content/plugins/wp-rocket/inc/main.php(50): WP_Rocket\Plugin->load()
#2 /var/www/example.com/htdocs/wp-includes/class-wp-hook.php(307): rocket_init()
#3 /var/www/example.com/htdocs/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()
#4 /var/www/example.com/htdocs/wp-includes/plugin.php(476): WP_Hook->do_action()
#5 /var/www/example.com/htdocs/wp-settings.php(461): do_action()
#6 /var/www/example.com/wp-config.php(123): require_once('/var/www/vasili...')
```

I believe it's coming from this:
https://github.com/wp-media/wp-rocket/blob/3a5884157156e4dfb4e37e5e91f4caf0e37f9ad7/inc/ThirdParty/ServiceProvider.php#L128

Can you please look into this? 🙏

**Review by vmanthos — APPROVED** [review]
This is working as expected. 👍 

**Note:**

This was tested with a "dummy" Polygon theme.
If the customer provides the theme, the PR will be retested.

**Pull request merged**